### ISSUE TITLE
MVELCondition does not re-throw exceptions

### ISSUE THREAD

**Issue by AleksandrPalchuk** [j-easy/easy-rules#211]
Hi

We have problem with identifying/handling errors on a condition execution. 
Exceptions in `MVELCondition` are suppressed and can't be handled in `onFailure(...)`  from `RuleListener`.

Code in `MVELCondition ` doesn't re-throw exceptions. The code example:

```
public boolean evaluate(Facts facts) {
        try {
            return (Boolean)MVEL.executeExpression(this.compiledExpression, facts.asMap());
        } catch (Exception var3) {
            LOGGER.error("Unable to evaluate expression: '" + this.expression + "' on facts: " + facts, var3);
            return false;
        }
    }
```

Similar issue was described here https://github.com/j-easy/easy-rules/issues/194 but in this time it is related to `MVELCondition`.  As I see, it was fixed already for `MVELAction`. 
Could you please help us in this situation? Thanks.

**Comment by zhhaojie** [comment]
it's hard to say , i had trying so hard to understand @benas 's design . 
so far, easy rule will throw exceptions when you define your rule ,and that time your rule being execute .

let's go back to `Rule` interface
```
    /**
     * Rule conditions abstraction : this method encapsulates the rule's conditions.
     * <strong>Implementations should handle any runtime exception and return true/false accordingly</strong>
     *
     * @return true if the rule should be applied given the provided facts, false otherwise
     */
    boolean evaluate(Facts facts);
```

if there is something wrong in your MVELCondition ,this one will NOT get the `execute` chance.  may be we should call them 'missed rule ' .

if `boolean evaluate(Facts facts) throw exception`, that's a very big change of easy-rules .

I am also very concerned about what @benas  will say.

**Comment by benas** [comment]
@zhhaojie You nailed it, thank you for the follow up! Your explanation is correct and this is what I had in mind.

@AleksandrPalchuk The contract of a rule clearly says that rule evaluation should not yield any exception and that implementations should handle errors. That's why there is no parameter of type `Exception` in `RuleListener#afterEvaluate`.

However, in the action, it is a different story. The contract of the rule allows you to throw an exception from the action and the listener allows you to get that exception. In #194, it was definitely a bug since I could throw an exception in my rule's action but I can't handle it in my listener. But that is fixed now in v3.3.

> Exceptions in MVELCondition are suppressed and can't be handled in onFailure(...) from RuleListener.

I see a small confusion here, let me tryand I will try to

**Comment by turiandras** [comment]
Hi,

I'd argue that we could further differentiate the condition-evaluates-to-false ('missed') case:
- condition MVEL is valid; expression 'compiles' and makes sense with the given facts, only the current facts does not satisfy the condition
- condition MVEL is invalid; there is a typo in the MVEL or maybe the data model of the facts has changed, rendering the previously valid condition invalid
 
I'd be happy if the 2 could be differentiated somehow. I understand that we don't want to throw exceptions from conditions; maybe we could somehow ask for the list of rules that had 'invalid' MVEL conditions after a rule engine execution, together with the exceptions thrown during their evaluation.

It would almost make the MVEL conditions type safe; we could learn immediately and exactly about changed Fact data model whenever we try the rules, instead of just silently not triggering the actions anymore.

I'm happy to attempt to implement it, but first I'm curious what do you think.

**Comment by Alexey1Gavrilov** [comment]
I do agree with @turiandras. A condition can be incorrectly evaluated to `false` due to an unexpected runtime exception which is swallowed in the MVELCondition

**Comment by benas** [comment]
The exception is not swallowed, it is logged as an error.

Practically speaking, what you would you do with the exception if it's re-thrown? 

> condition MVEL is invalid; there is a typo in the MVEL

Let's be pragmatic here, when there is a typo in the expression, you will see it in the log and fix it. And IMHO, if the condition is correctly unit-tested upfront, you should not end-up in this situation in the first place.

**Comment by Alexey1Gavrilov** [comment]
I’d like to fail the processing with internal server error if the condition suddenly starts throwing an exception if, for example, a database is temporary not available. 
I agree that the logged error can be noticed during testing but since it won’t cause a test failure it may actually be missed causing unexpected behavior in prod.

**Comment by benas** [comment]
I might be wrong, so let me re-open this for further consideration.

There was an attempt to implement this in the past (See 2505d8ac22625c78d8d7fa19b23ddfdfeaa32b01 and #112), but to be honest, I don't remember the reason why it was abandoned. IIRC from the commit message, there was no way to distinguish between a missing fact, an incorrect injected fact type and an error in the expression.

We can re-throw runtime exceptions in `MVELCondition#evaluate` and `SpELCondition#evaluate` or let them bubble up, but what should the engine do with them?

Here is my second attempt: 962b8ecb0a30ba0b98b89a3bb4048c870a74edf2. It lets runtime exceptions in MVEL/SpEL conditions bubble up and handles them at the engine level. The engine then logs the error and gives the option to stop or continue with next rules (using the `skipOnFirstNonTriggeredRule ` parameter).

I would not change the listener and call it since those are runtime exceptions and not part of the contract (as it is the case with the `execute` method).

@AleksandrPalchuk @turiandras @Alexey1Gavrilov @zhhaojie What do you think?

**Comment by benas** [comment]
I did some research on this and tried to compare things with drools. It turns out that drools behaviour is unpredictable regarding errors during the evaluation of the condition. Here is an example based on the [official hello world](https://github.com/kiegroup/drools/blob/master/drools-examples/src/main/resources/org/drools/examples/helloworld/HelloWorld.drl) sample:

1. Domain object `Message`:

```java
public class Message {
    private String type;
    // getter and setter for type
    public void printMessage() {
        System.out.println("Type: " + type);
    }
}
```

2. Rule defined in `rules.drl`:

```yml
import com.example.demo.Message

rule "Hello World"
dialect "mvel"
when
  message:Message (type == 'Hello')
then
  message.printMessage();
end
```

3. Code to run the sample:

```java
public class DemoDrools {

	public static void main(String[] args) throws Exception {
		PackageBuilder packageBuilder = new PackageBuilder();
		packageBuilder.addPackageFromDrl(new FileReader("rules.drl"));
		RuleBase ruleBase = RuleBaseFactory.newRuleBase();
		ruleBase.addPackage(packageBuilder.getPackage());
		WorkingMemory workingMemory = ruleBase.newStatefulSession();
		Message message = new Message();
		message.setType("Hello");
		workingMemory.insert(message);
		workingMemory.fireAllRules(); // prints Type: Hello
	}

}
```

If someone wants to play with the sample and run the following tests, the complete code can be found here:  [demo-drools.zip](https://github.com/j-easy/easy-rules/files/3924055/demo-drools.zip)

Now, here are my (monkey) tests:

1. Test 1: Typo in the condition

```diff
when
-  message:Message (type == 'Hello')
+  message:Message (typex == 'Hello')
then
```

This fails with `org.drools.core.RuntimeDroolsException: Field/method 'typex' not found for class 'com.example.demo.Message'`

2. Test 2: Incorrect condition v1

```diff
when
-  message:Message (type == 'Hello')
+  message:Message (type > 'Hello')
then
```

Here the rule is not fired. This is equivalent to considering the condition as evaluated to false. Same behaviour as Easy Rules.

3. Test 3: Incorrect condition v2

```diff
when
-  message:Message (type == 'Hello')
+  message:Message (type === 'Hello')
then
```

Fails with NPE...

So in hindsight, I believe Easy Rules should not make any decision in the case of a runtime exception while evaluating the condition. It should simply let the exception bubble up from the condition and stop rule processing at the engine level (this will basically fix both issues described by @Alexey1Gavrilov [here](https://github.com/j-easy/easy-rules/issues/211#issuecomment-559574889)).

Wdyt? Looking forward for feedback to find the best way to fix this issue .

**Comment by turiandras** [comment]
By 'stop rule processing at the engine level', do you mean the behavior implemented in 962b8ec ?

If I understand it correctly, the behaviour of easy-rules doesn't really change from the viewpoint of the application. If there is a typo in the condition mvel, or there is a temporary runtime problem making some subset of the facts unavailable/null/invalid (good point @Alexey1Gavrilov ), rules will be treated the same as if they would be simply 'missed', and not as a special problem case.

It is true that the problem would be logged, and unit tests could catch the lack of action for typos.

However, here are some counterpoint why I feel this might be insufficient:
- in my experience, in a big application the production logs are not always routinely read proactively; the 1st/2nd/3rd level support reads it when they learn about a problem from the users. Ideally, ERROR level logs could be monitored proactively, I don't know how frequently does that actually happen in practice.
- for the tests to cover this, they have to be extensive. Eg, for a condition ( aa.bb == 12 || cc.dd == 8 ), both OR-ed components of the rule needs a specific test case just to make sure that the intended component alone causes the rule's action to happen. So we factored out the logic to a rules engine, but we still have to write java code covering every condition component in the unit tests to protect against typos.
- we could say that the conditions have to be written once, they are tested manually, logs can be checked during development, we could make sure there are no typos by being careful and diligent.
However, a good, working condition can become a typo when the underlying data-model of the Fact class hierarchies change. If the Facts and the rules are maintained by the same person, this shouldn't be a big problem, but with big applications and team(s) of developers, this could become painful.
- Alexey's point (as I understand it): some of the facts can come from an external system, a DB or something else. If there is a problem with that at runtime, it is possible that some of the facts graph will be empty. The new RuntimeException in RuleProxy in 962b8ec might take care of this when NoSuchFactException is thrown, I'm not sure if that covers objects missing deep down in the fact object hierarchy or only the case when some of the top level facts are missing.
If it covers both, then ignore the rest of this point. In case objects deep in the fact graph (eg user.favourites.movies.size() < 3 ) being null does not result in NoSuchFactException, we would just act like the data is available but it doesn't satisfy the condition (except for the logs of course). You could argue that checking availability of external systems is not the job of easy rules. That is true, but still it could be helpful to distinguish missing data from non-triggering data. A workaround for this can be to have rules that raise errors if the data is missing.


What I would find ideal could be either of these:
- onEvaluationFailure listener method - the application can decide if it wants to return an error message, after potentially examining the cause of the evaluation failure
- after the engine execution, having a list of rules-exception pairs available to the application, describing all the rules that had problems evaluating the conditions

Both approaches would enable the application to potentially throw an application exception when any of the conditions are invalid. It would protect against typos and fact model changes without a high number of unit test cases, almost making the mvel conditions statically typed.
And also the application could decide how to handle rules whose conditions fail because of missing data; in some cases simply not triggering the action is already sufficient, in other cases it could show a warning or an error.

Sorry for the long text, and thanks for the library btw! :)

**Comment by benas** [comment]
Thank you for your feedback! Much appreciated.

> By 'stop rule processing at the engine level', do you mean the behavior implemented in 962b8ec ?

Yes, since the MVEL/SpEL conditions now throw exceptions (instead of catching them and return false), the engine gives the option to either:

* skip next rules processing (if you set the `SkipOnFirstNonTriggeredRule` parameter)
* continue with next rules by considering the condition as if it returned false (current behaviour)

The notable change here is that any test will now fail if there is typo in the condition or a breaking change in the model (you see that I even removed the logging at the condition level).

In regards to your bullet points, it is that simple: It is not Easy Rules' role to try to distinguish the failure reason (if it is a typo in the condition, or an incorrect expression, or a missing field in the model, etc). As far as Easy Rules is concerned, the condition evaluation should succeed (return true or false) or fail with a runtime exception for whatever reason (the `NoSuchFactException` is really a convenience to give a nice log message, nothing more).

In regards to the listener, what would be the signature of the new/changed method? Here are the options I see:

#### Option 1: Add a new method `onEvaluateFailure`

```diff
void afterEvaluate(Rule rule, Facts facts, boolean evaluationResult);
+void onEvaluateFailure(Rule rule, Facts facts, Exception exception);
```

This is ugly from an API design point of view. As a user, I have to implement two methods for the failure case (since evaluationResult is a primitive boolean, I need to check if exception is null to distinguish between an evaluation that succeeded and returned false and an evaluation that failed due to an exception)

#### Option 2: Add a new parameter of type Throwable/Exception to `afterEvaluate`

```diff
-void afterEvaluate(Rule rule, Facts facts, boolean evaluationResult);
+void afterEvaluate(Rule rule, Facts facts, boolean evaluationResult, Exception exception);
```

This is ugly too. If evaluationResult=false, I need to check if the exception is null to know if the evaluation failed due to an exception or if it succeeded and the condition returned false.

#### Option 3: Split the afterEvaluate in two methods

```diff
- void afterEvaluate(Rule rule, Facts facts, boolean evaluationResult);
+ void onEvaluateSuccess(Rule rule, Facts fact);
+ void onEvaluateFailure(Rule rule, Facts facts, Exception exception);
```

This is also ugly. It is an attempt to make it like `onSuccess/onFailure` methods for actions but it's not appropriate for conditions. Condition evaluation success can yield to either true or false, which cannot be captured with this option.

#### Option 4: encapsulate the boolean + exception in a EvaluationResult concept

No, encapsulation is good, but not for this case..

#### Option 5: Open a JEP to suggest making Java methods return multiple values: The boolean + the exception if any

Well, why not.

The difficulty is that a condition method returns a boolean and might also throw a runtime exception. So the question is what should the engine do when the condition throws an exception and what would the listener method signature be. More precisely, we need to do the following TODO (🙃) in the rules engine' code:

```java
boolean evaluationResult = false;
try {
    evaluationResult = rule.evaluate(facts);
} catch (RuntimeException exception) { // ignoring runtime exceptions and crash the whole thing a la drools is not an option
    LOGGER.error("Rule '" + name + "' evaluated with error", exception);
    // TODO call the listener, but with which signature? And what if the listener method throws a runtime exception..
    // give the option to skip next rules on evaluation failure:
    if (parameters.isSkipOnFirstNonTriggeredRule()) {
        LOGGER.debug("Next rules will be skipped since parameter skipOnFirstNonTriggeredRule is set");
        break;
    }
    // otherwise, consider the evaluation as false and continue with next rules
}
```

**Comment by zhhaojie** [comment]
sorry for late, according to design,  MVEL and SpEL would throw `CompileException`,`ExpressionException`,`EvaluationException`.

I insist that how to ensure that the rules are written correctly is **not** the engine
Responsibility. the MVEL, SpEL is. it is impossible to take care of all *EL's Exception.
image we use SQL to write rule? How much trouble we will meet? in this case, what Exception should be throw while `evaluate`?  what about a wrapException?

i would like to define `RuleListener` like (ugly but simple) :
```
boolean beforeEvaluate(Rule rule, Facts facts);
boolean onEvaluateSuccess(Rule rule, Facts facts);
boolean onEvaluateFailure(Rule rule, Facts facts, Exception exception);
void beforeExecute(Rule rule, Facts facts);
void onExecuteSuccess(Rule rule, Facts facts);
void onExecuteFailure(Rule rule, Facts facts, Exception exception);
```

i had been use `easy-rules` in product, it works fine. thank you all

**Comment by benas** [comment]
@zhhaojie 

> I would like to define RuleListener like (ugly but simple) :

> boolean beforeEvaluate(Rule rule, Facts facts);
> void onEvaluateSuccess(Rule rule, Facts facts);
> void onEvaluateFailure(Rule rule, Facts facts, Exception exception);

This is basically option 3 in my [previous comment](https://github.com/j-easy/easy-rules/issues/211#issuecomment-567262842), which as mentioned, does not allow to distinguish between an evaluation that succeeded and returned true and an evaluation that succeeded and returned false since there is no `evaluationResult` parameter in `onEvaluateSuccess(Rule rule, Facts facts)`. Do you see?

Please note that a condition that evaluates to `false` is **not** a failure, and the engine should **not** call `onEvaluateFailure(Rule rule, Facts facts, Exception exception)` with a `null` for the third parameter `exception`. When a condition is correctly evaluated to either `true` or `false` (aka no typos, nor runtime exceptions, etc), then this a *successful* evaluation and the engine should call `onEvaluateSuccess(Rule rule, Facts facts)`, but in this case, in my listener I can't know if my condition evaluated to true or false using this signature (since there is no `evaluationResult` parameter). I hope I made things more clear now and why this option is not a good one IMO.

Using onEvaluateSuccess/onEvaluateFailure would be really confusing, because it would be easy to make the mistake of thinking that:

* onEvaluateSuccess -> condition returned true
* onEvaluateFailure -> condition returned false

which is incorrect as described previously. If we really want to have a method in the listener to get notified about runtime exceptions while evaluating a condition, we can add a new method with a more explicit name like `onEvaluationError` (not the word error, not failure). So we would have something like:

```java
void afterEvaluate(Rule rule, Facts facts, boolean evaluationResult);
void onEvaluationError(Rule rule, Facts facts, Exception exception);
```

This is basically option 1, but with a better name (naming is hard ..). I said this option is ugly because "As a user, I have to implement two methods for the failure case". This is actually not correct because for the "failure" case, I have to implement only one method `onEvaluationError`. For the evaluation success case, I also implement only one method `afterEvaluate` which gives me the ability to know if the evaluation was true or false thanks to the `evaluationResult` parameter. So I think this option is the best one.

Now to sum up, here is what is missing today (as of v3.4) as reported by @AleksandrPalchuk , @turiandras and @Alexey1Gavrilov :

1. (MVEL) Conditions do not throw exceptions (any runtime exception is only logged and the condition returns false), so it is impossible to detect and distinguish evaluation error/result in a unit test
2. There is no way to intercept evaluation error in RuleListener
3. There is no way to stop the engine on a rule evaluation runtime error

I fixed that in 77c616e:

1. MVEL/SpEL condition now throw any runtime exceptions, so it is possible to detect any failures due to incorrect expressions early in unit tests.
2. It is possible to get notified about those runtime exception in `RuleListener#onEvaluationError`
3. It is possible to choose between stopping the engine on rule evaluation runtime error (using the `isSkipOnFirstNonTriggeredRule` parameter) or consider the evaluation as false and continue (current behaviour)

Version 4.0.0-SNAPSHOT contains the fix and has been [deployed](https://github.com/j-easy/easy-rules/wiki/getting-started#use-a-snapshot-version) to sonatype's snapshot repo. I would be grateful to anyone who could give it a try and share feedback.

Finally, I strongly agree with @zhhaojie on the following:

> I insist that how to ensure that the rules are written correctly is not the engine
Responsibility. the MVEL, SpEL is. it is impossible to take care of all *EL's Exception.

It is not ER concern to deal with runtime exceptions for EL expressions in both conditions and actions. It is just impossible because there are many differences in behaviour between EL providers. For example, MVEL [throws](https://github.com/j-easy/easy-rules/blob/master/easy-rules-mvel/src/test/java/org/jeasy/rules/mvel/MVELConditionTest.java#L49) an exception when a fact is missing, while SpEL silently [ignores](https://github.com/j-easy/easy-rules/blob/master/easy-rules-spel/src/test/java/org/jeasy/rules/spel/SpELConditionTest.java#L54) a missing fact and returns false. The user should be aware of those differences and it is up to him/her to choose which EL provider to use. As far as the rules engine is concerned, it should try to evaluate the condition. If the condition returns true or false, that's great and it can execute/skip the action accordingly. If the condition has failed to be evaluated for any reason (typo, missing fact, incorrect fact type, etc), the engine should:

1. log the error
2. give the opportunity to the user to do something with the error using an extension point (here, it is the newly added `RuleListener#onEvaluationError`)
3. give the user the choice to stop immediately or consider the evaluation as false and continue with subsequent rules.

This is what actually has been implemented in 77c616e and will be clearly documented.

Sorry for the long message, but when it comes to API design, I tend to be more careful now (especially about naming, which is really hard..)

**Comment by benas** [comment]
I'm closing this issue as I believe it has been solved. Thanks to @AleksandrPalchuk for opening the issue and for everyone who has participated in the discussion!
